smac/intensification/intensification.py

```diff
@@ -9,33 +8,89 @@
 from smac.stats.stats import Stats
 from smac.utils.constants import MAXINT
 from smac.configspace import Configuration
-from smac.runhistory.runhistory import RunHistory, InstSeedBudgetKey
-from smac.tae.execute_ta_run import BudgetExhaustedException, CappedRunException, ExecuteTARun
+from smac.runhistory.runhistory import (
+    InstSeedBudgetKey,
+    RunInfo,
+    RunHistory,
+    RunValue,
+    StatusType
+)
 from smac.utils.io.traj_logging import TrajLogger
-from smac.intensification.abstract_racer import AbstractRacer, _config_to_run_type
+from smac.intensification.abstract_racer import (
+    AbstractRacer,
+    IntensifierBehest,
+    _config_to_run_type,
+)
 from smac.optimizer.epm_configuration_chooser import EPMChooser
 
 __author__ = "Katharina Eggensperger, Marius Lindauer"
 __copyright__ = "Copyright 2018, ML4AAD"
 __license__ = "3-clause BSD"
 
 
+class NoMoreChallengers(Exception):
+    """Indicates that no more challengers are available
+    for the intensification to proceed"""
+    pass
+
+
 class IntensifierStage(Enum):
     """Class to define different stages of intensifier"""
     RUN_FIRST_CONFIG = 0  # to replicate the old initial design
     RUN_INCUMBENT = 1  # Lines 3-7
     RUN_CHALLENGER = 2  # Lines 8-17
     RUN_BASIS = 3
 
+    # helpers to determine what type of run to process
+    # A challenger is assumed to be processed if the stage
+    # is not from first_config or incumbent
+    PROCESS_FIRST_CONFIG_RUN = 4
+    PROCESS_INCUMBENT_RUN = 5
+
 
 class Intensifier(AbstractRacer):
-    """Races challengers against an incumbent (a.k.a. SMAC's intensification
-    procedure).
+    """Races challengers against an incumbent
+
+    SMAC's intensification procedure is denoted below.
+    This class honors bellow procedure and implements it's sequential
+    nature via stages (IntensifierStage), which are also highlighted below
+
+    Procedure 2: Intensify(Θ~new, θinc, M, R, tintensify, Π, cˆ)
+    cˆ(θ, Π0) denotes the empirical cost of θ on the subset of instances
+    Π0 ⊆ Π, based on the runs in R; maxR is a parameter
+    ____________________________________________________________
+    1 for i := 1, . . . , length(Θ~ new) do
+    2 θnew ← Θ~ new[i]
```

Comment:
Would it be possible to use indents here as in SMAC's procedure 2?

Reply:
Yes, I added proper indentation now